src/qiboml/models/abstract.py

```diff
@@ -0,0 +1,40 @@
+"""Defines the general structure of a qiboml module"""
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+from qibo import Circuit
+from qibo.config import raise_error
+from qibo.gates import abstract
+
+from qiboml.backends import JaxBackend
+
+
+@dataclass
+class QuantumCircuitLayer(ABC):
+
+    nqubits: int
+    qubits: list[int] = None
+    circuit: Circuit = None
+    initial_state: "ndarray" = None
+    backend: "qibo.backends.Backend" = JaxBackend()
+
+    def __post_init__(self) -> None:
+        if self.qubits is None:
+            self.qubits = list(range(self.nqubits))
```

Comment:
The problem I see here, though, is that then you will need to construct the layer as:
```
QuantumCircuitLayer(nqubits=5, _qubits=[0,1,2])
```
if you want to specify the qubits. I am not sure if there's any workaround with `dataclasses`.

Reply:
Interestingly enough, you can do it with Pydantic:
```py
class Ciao(BaseModel):
    name_: str = Field(alias="name")
                                                                                                                                                                                      
    @property
    def name(self):
        print(f"Ciao, {self.name_}")
        return self.name_
                                                                                                                                                                                      
c = Ciao(name="come va?")
c.name
```
(though you can not use `._qubit`, since fields are supposed to be public, and they can not contain leading underscores)